explain-joins.md

```diff
@@ -0,0 +1,267 @@
+
+An example Hash Join is as follows:
+
+{{< copyable "sql" >}}
+
+```sql
+EXPLAIN SELECT /*+ HASH_JOIN(t1, t2) */ * FROM t1, t2 WHERE t1.id = t2.id;
+```
+
+```sql
++-----------------------------+-----------+-----------+---------------+------------------------------------------------+
+| id                          | estRows   | task      | access object | operator info                                  |
++-----------------------------+-----------+-----------+---------------+------------------------------------------------+
+| HashJoin_27                 | 142020.00 | root      |               | inner join, equal:[eq(test.t1.id, test.t2.id)] |
+| ├─TableReader_29(Build)     | 142020.00 | root      |               | data:TableFullScan_28                          |
+| │ └─TableFullScan_28        | 142020.00 | cop[tikv] | table:t1      | keep order:false                               |
+| └─TableReader_31(Probe)     | 180000.00 | root      |               | data:TableFullScan_30                          |
+|   └─TableFullScan_30        | 180000.00 | cop[tikv] | table:t2      | keep order:false                               |
++-----------------------------+-----------+-----------+---------------+------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+The execution process of `Hash Join` is as follows:
+
+1. Cache the data of the `Build` side in memory.
+2. Construct a Hash Table on the `Build` side based on the cached data.
+3. Read the data at the `Probe` side.
+4. Use the data of the `Probe` side to probe the Hash Table.
+5. Return qualified data to the user.
+
+The operator info column in the `explain` table also records other information about `Hash Join`, including whether the query is Inner Join or Outer Join, and what are the conditions of Join. In the above example, the query is an Inner Join, where the Join condition `equal:[eq(test.t1.id, test.t2.id)]` partly corresponds with the query statement `WHERE t1.id = t2.id`. The operator info of the other Join operators in the following examples is similar to this one.
+
+### Runtime Statistics
+
+If [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) (default: 1GB) is exceeded, TiDB will attempt to use temporary storage provided that `oom-use-tmp-storage=TRUE` (default). This means that the hash table used as part of the hash join is created on disk. Runtime statistics, such as memory usage are visible in the `execution info` of `EXPLAIN ANALYZE`. The following example shows the output of `EXPLAIN ANALYZE` with a 1GB (default) `tidb_mem_quota_query` quota, and a 500MB quota. At 500MB the hash table spills to disk:
```

Comment:
> This means that the hash table used as part of the hash join is created on disk.

@SunRunAway please confirm: is the build side data cached in memory spilled to disk or the hash table itself is spilled to disk?

Reply:
I think we should update https://docs.pingcap.com/tidb/stable/system-variables#tidb_mem_quota_query as well so its a bit clearer what this does.

Reply:
For now, it is that build side data cached in memory spilled to disk, but we should also consider the spilling for hash table in the future.

Reply:
It's better to clearly describe the current behavior on the document. we can update the document once we spilled hash table to disk.

BTW, for technical reasons, it's not recommended to spill hash table to disk. There should be other methods to achieve the same goal.

Reply:
I have attempted to clarify this in https://github.com/pingcap/docs/pull/3922/commits/bcb16dd1551c17a9e16abd5fefa027cbd4505778 - PTAL, thx :-)